src/webgpu/api/validation/attachment_compatibility.spec.ts

```diff
@@ -0,0 +1,305 @@
+            pass.executeBundles([bundle]);
+            pass.endPass();
+            return encoder.finish();
+          },
+        };
+      }
+      case 'render pass':
+        return {
+          encoder: pass,
+          finish() {
+            pass.endPass();
+            return encoder.finish();
+          },
+        };
+    }
+  }
+
+  createRenderPipeline(
+    colorStates: Iterable<GPUColorStateDescriptor>,
+    depthStencilState?: GPUDepthStencilStateDescriptor
+  ) {
+    const wgslVertex = `
+    [[builtin(position)]] var<out> position : vec4<f32>;
+
+    [[stage(vertex)]]
+    fn main() -> void {
+      position = vec4<f32>(0.0, 0.0, 0.0, 0.0);
+      return;
+    }
+  `;
+    const wgslFragment = `
+    [[stage(fragment)]]
+    fn main() -> void {
+      return;
+    }
+  `;
+
+    return this.device.createRenderPipeline({
+      vertexStage: {
+        module: this.device.createShaderModule({
+          code: wgslVertex,
+        }),
+        entryPoint: 'main',
+      },
+      fragmentStage: {
+        module: this.device.createShaderModule({
+          code: wgslFragment,
+        }),
+        entryPoint: 'main',
+      },
+      primitiveTopology: 'triangle-list',
+      colorStates,
+      depthStencilState,
+    });
+  }
+}
+
+export const g = makeTestGroup(F);
+
+const kColorAttachmentFormats = kRegularTextureFormats
+  .map((key): [RegularTextureFormat, typeof kRegularTextureFormatInfo[RegularTextureFormat]] => [
+    key,
+    kRegularTextureFormatInfo[key],
+  ])
+  .filter(([, format]) => format.color && format.renderable)
+  .map(([key]) => key);
+
+g.test('render_pass_and_bundle_color_format')
+  .desc('Test that color attachment formats in render passes and bundles must match.')
+  .params(
+    params()
+      .combine(poptions('passFormat', kColorAttachmentFormats))
+      .combine(poptions('bundleFormat', kColorAttachmentFormats))
+  )
+  .fn(t => {
+    const { passFormat, bundleFormat } = t.params;
+    const bundleEncoder = t.device.createRenderBundleEncoder({
+      colorFormats: [bundleFormat],
+    });
+    const bundle = bundleEncoder.finish();
+    const encoder = t.device.createCommandEncoder();
+    const pass = encoder.beginRenderPass({
+      colorAttachments: [t.createColorAttachment(passFormat)],
+    });
+    pass.executeBundles([bundle]);
+    pass.endPass();
+    t.expectValidationError(() => {
+      t.queue.submit([encoder.finish()]);
+    }, passFormat !== bundleFormat);
+  });
+
+g.test('render_pass_and_bundle_color_count')
+  .desc('Test that the number of color attachments in render passes and bundles must match.')
+  .params(
+    params()
+      .combine(poptions('passCount', kColorAttachmentCounts))
+      .combine(poptions('bundleCount', kColorAttachmentCounts))
+  )
+  .fn(t => {
+    const { passCount, bundleCount } = t.params;
+    const bundleEncoder = t.device.createRenderBundleEncoder({
+      colorFormats: range(bundleCount, () => 'rgba8unorm'),
```

Comment:
It would be nice to test varying formats in postitions other than 0 too, but maybe that can be done with the sparse attachments.

Reply:
Ditto below

Reply:
I'm concerned about the test case count explosion. `kColorAttachmentFormats` currently has 31 formats. The `color_format` tests currently create 961 and 1922 test cases (`render_pass_and_bundle` and  `render_pass_or_bundle_and_pipeline` respectively). Combining `color_format` and `color_count` would create a combined tests with 15376 and 30752 tests cases.

Not sure that sparse attachments would create any less test cases.

Reply:
I don't think we need to test every combination, just some. So e.g. we could just test:

- pass [rgba8unorm, rgba8unorm, rgba8unorm, ...] vs bundle with same array except one element changed to bgra8unorm or something (+ `kMaxColorAttachments` cases)
- pass [fmt1, fmt1, fmt1, ...] vs bundle with same array except element `[2]` changed to each color format (+ `kColorAttachmentFormats.length` cases)

@kangz is that what you had in mind?

And yeah I definitely think it's fine to do this in the PR where we update for sparse attachments, or as a separate PR before or after that